File: app/src/main/java/com/hop/pirate/activity/MainActivity.java
package com.hop.pirate.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.widget.TextView;

import com.hop.pirate.Constants;
import com.hop.pirate.R;
import com.hop.pirate.base.BaseActivity;
import com.hop.pirate.callback.AlertDialogOkCallBack;
import com.hop.pirate.callback.ResultCallBack;
import com.hop.pirate.event.EventClearAllRequest;
import com.hop.pirate.event.EventLoadWalletSuccess;
import com.hop.pirate.event.EventRechargeSuccess;
import com.hop.pirate.event.EventSkipTabPacketsMarket;
import com.hop.pirate.event.EventSyncVersion;
import com.hop.pirate.fragment.TabHomeFragment;
import com.hop.pirate.fragment.TabPacketsMarketFragment;
import com.hop.pirate.fragment.TabWalletFragment;
import com.hop.pirate.model.MainModel;
import com.hop.pirate.model.bean.WalletBean;
import com.hop.pirate.model.impl.MainModelImpl;
import com.hop.pirate.service.HopService;
import com.hop.pirate.service.WalletWrapper;
import com.hop.pirate.util.BottomNavigatorAdapter;
import com.hop.pirate.util.FragmentNavigator;
import com.hop.pirate.util.Utils;
import com.hop.pirate.view.BottomNavigatorView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import androidLib.AndroidLib;

public class MainActivity extends BaseActivity implements androidLib.HopDelegate {
    public static WalletBean sWalletBean;
    public static boolean isSyncVersion = false;
    private MainModel mMainModel;
    public static final String TAG = "HopProtocol";
    public static final int ATSysSettingChanged = 1;
    public static final int ATPoolsInMarketChanged = 2;
    public static final int ATCounterDataRead = 3;
    public static final int ATNeedToRecharge = 4;
    public static final int ATRechargeSuccess = 5;
    private Class[] mFragmentArray = {TabHomeFragment.class, TabPacketsMarketFragment.class,
            TabWalletFragment.class};
    private BottomNavigatorView bottomNavigatorView;
    private FragmentNavigator mNavigator;
    private TextView mGetFreeCoinTv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initNavigator(savedInstanceState);
        setContentView(R.layout.activity_main);
        EventBus.getDefault().register(this);
        mMainModel = new MainModelImpl();
        if (Utils.getApplication(this).isRunning()) {
            return;
        }
        initService();
    }


    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        if (Utils.getApplication(this).isRunning()) {
            return;
        } else {
            AndroidLib.stopProtocol();
        }
        initService();


    }

    void initService() {
        mMainModel.initService(this, this, new ResultCallBack<String>() {
            @Override
            public void onError(Throwable e) {
                Utils.showOkAlert(MainActivity.this, R.string.tips, R.string.blockchain_sync_error, new AlertDialogOkCallBack() {

                    @Override
                    public void onClickOkButton(String parameter) {
                        finish();
                    }
                });

            }

            @Override
            public void onSuccess(String str) {

            }

            @Override
            public void onComplete() {
                loadWallet(false);
                mMainModel.syncVersion();
            }
        });

    }


    public void loadWallet(final boolean isShowLoading) {
        mMainModel.getWalletInfo(this, new ResultCallBack<WalletBean>() {
            @Override
            public void onError(Throwable e) {
                if (isShowLoading) {
                    dismissDialogFragment();
                }
                Utils.toastException(MainActivity.this, e, Constants.REQUEST_WALLET_INFO_ERROR);
            }

            @Override
            public void onSuccess(WalletBean walletBean) {
                sWalletBean = walletBean;
                WalletWrapper.MainAddress = walletBean.getMain();
                WalletWrapper.SubAddress = walletBean.getSub();
                WalletWrapper.EthBalance = walletBean.getEth();
                WalletWrapper.HopBalance = walletBean.getHop();
                WalletWrapper.Approved = walletBean.getApproved();
                String tokenName = Utils.getString(Constants.CUR_SYMBOL, "HOP");
                if (sWalletBean.getHop() != 0 && tokenName.startsWith("HOP")) {
                    gone(mGetFreeCoinTv);
                }

            }

            @Override
            public void onComplete() {
                if (isShowLoading) {
                    dismissDialogFragment();
                }
                EventBus.getDefault().postSticky(new EventLoadWalletSuccess());
            }
        });
    }

    private void initNavigator(Bundle savedInstanceState) {
        BottomNavigatorAdapter bottomNavigatorAdapter = new BottomNavigatorAdapter(this);
        for (Class fragment : mFragmentArray) {
            bottomNavigatorAdapter
                    .addTab(new BottomNavigatorAdapter.TabInfo(fragment.getSimpleName(), fragment, null));
        }
        mNavigator = new FragmentNavigator(getSupportFragmentManager(), bottomNavigatorAdapter, R.id.contentFrame);
        mNavigator.setDefaultPosition(Constants.TAB_HOME);
        mNavigator.onCreate(savedInstanceState);
    }

    @Override
    public void initViews() {
        bottomNavigatorView = findViewById(R.id.main_bottom_navigator_view);
        mGetFreeCoinTv = findViewById(R.id.getFreeCoinTv);
        setCurrentTab(mNavigator.getCurrentPosition());

        bottomNavigatorView.setOnBottomNavigatorViewItemClickListener(new BottomNavigatorView.OnBottomNavigatorViewItemClickListener() {
            @Override
            public void onBottomNavigatorViewItemClick(int position, View view) {
                setCurrentTab(position);
            }
        });
        mGetFreeCoinTv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setCurrentTab(Constants.TAB_WALLET);
                gone(mGetFreeCoinTv);
            }
        });
    }

    @Override
    public void initData() {
    }

    private void setCurrentTab(int position) {
        if (null != mNavigator) {
            mNavigator.showFragment(position);
        }
        if (null != bottomNavigatorView) {
            bottomNavigatorView.select(position);
        }
        setFreeCoinStatus(position);
    }

    private void setFreeCoinStatus(int position) {
        if (position != Constants.TAB_WALLET && WalletWrapper.EthBalance == 0) {
            visiable(mGetFreeCoinTv);
        } else {
            gone(mGetFreeCoinTv);
        }
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    public void addMiningPool(EventSkipTabPacketsMarket eventSkipTabPacketsMarket) {
        setCurrentTab(Constants.TAB_RECHARGE);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void eventSyncVersion(EventSyncVersion eventSyncVersion) {
        mMainModel.syncVersion();
    }


    @Override
    protected void onSaveInstanceState(final Bundle outState) {
        super.onSaveInstanceState(outState);
        mNavigator.onSaveInstanceState(outState);
    }


    @Override
    public void actionNotify(short type, String msg) {
        Log.w(TAG, "actionNotify: type:[" + type + "] msg:=>" + msg);
        switch ((int) type) {
            case ATSysSettingChanged:
                break;
            case ATPoolsInMarketChanged:
                break;
            case ATNeedToRecharge:
                Utils.toastTips(getResources().getString(R.string.packets_insufficient_need_recharge));
                break;
            case ATRechargeSuccess:
                EventBus.getDefault().post(new EventRechargeSuccess());
                break;
            default:
                break;

        }
    }


    @Override
    public void serviceExit(Exception err) {
        //Broad cast to restart service
        Utils.getApplication(this).setRunning(false);
        HopService.stop();

    }

    @Override
    public void log(String str) {
        Log.i(TAG, str);
    }

    @Override
    public void cancelWaitDialog() {
        super.cancelWaitDialog();
        mMainModel.removeAllSubscribe();
        EventBus.getDefault().post(new EventClearAllRequest());
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void rechargeSuccess(EventRechargeSuccess eventRechargeSuccess) {
        loadWallet(false);
    }


    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            Intent intent = new Intent(Intent.ACTION_MAIN);
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            intent.addCategory(Intent.CATEGORY_HOME);
            startActivity(intent);
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        WalletWrapper.closeWallet();
        if (!Utils.getApplication(this).isRunning()) {
            AndroidLib.stopProtocol();
        }
        if (mMainModel != null) {
            mMainModel.removeAllSubscribe();
        }

        EventBus.getDefault().unregister(this);
    }


}

File: app/src/main/java/com/hop/pirate/ui/activity/MainActivity.kt
package com.hop.pirate.ui.activity

import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import androidLib.AndroidLib
import androidx.lifecycle.Observer
import com.hop.pirate.BR
import com.hop.pirate.Constants
import com.hop.pirate.HopApplication
import com.hop.pirate.R
import com.hop.pirate.databinding.ActivityMinePoolBinding
import com.hop.pirate.event.EventReLoadWallet
import com.hop.pirate.event.EventRechargeSuccess
import com.hop.pirate.event.EventShowHomeTip
import com.hop.pirate.event.EventSkipTabPacketsMarket
import com.hop.pirate.guide.PirateOnTopPosCallback
import com.hop.pirate.model.bean.WalletBean
import com.hop.pirate.service.WalletWrapper
import com.hop.pirate.ui.fragement.TabHomeFragment
import com.hop.pirate.ui.fragement.TabPacketsMarketFragment
import com.hop.pirate.ui.fragement.TabWalletFragment
import com.hop.pirate.util.Utils
import com.hop.pirate.viewmodel.MainVM
import com.hop.pirate.widget.navigator.BottomNavigatorAdapter
import com.hop.pirate.widget.navigator.BottomNavigatorView.OnBottomNavigatorViewItemClickListener
import com.hop.pirate.widget.navigator.FragmentNavigator
import com.nbs.android.lib.base.BaseActivity
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import zhy.com.highlight.HighLight
import zhy.com.highlight.shape.RectLightShape

class MainActivity : BaseActivity<MainVM, ActivityMinePoolBinding>() {

    companion object {
        var walletBean: WalletBean? = null
        const val TAG = "HopProtocol"
        const val ATSysSettingChanged = 1
        const val ATPoolsInMarketChanged = 2
        const val ATCounterDataRead = 3
        const val ATNeedToRecharge = 4
        const val ATRechargeSuccess = 5

        var firstCreateAccountActivity: Boolean = true
    }

    lateinit var higghtLight: HighLight
    var netTyp = Constants.DEFAULT_MAIN_NET

    private val mFragmentArray = arrayOf(TabHomeFragment::class.java, TabPacketsMarketFragment::class.java, TabWalletFragment::class.java)
    private lateinit var mNavigator: FragmentNavigator

    override fun onContentViewBefor(savedInstanceState: Bundle?) {
        super.onContentViewBefor(savedInstanceState)
        initNavigator(savedInstanceState)
    }

    override fun getLayoutId(): Int = R.layout.activity_main

    override fun initView() {
        netTyp = Utils.getInt(Constants.NET_TYPE, Constants.DEFAULT_MAIN_NET)
        EventBus.getDefault().register(this)
        if (HopApplication.instance.isRunning) {
            return
        }

        mViewModel.getWalletInfo(false)
        main_bottom_navigator_view.setOnBottomNavigatorViewItemClickListener(object : OnBottomNavigatorViewItemClickListener {
            override fun onBottomNavigatorViewItemClick(position: Int, view: View?) {
                setCurrentTab(position)
            }
        })
        get_free_coin_tv.setOnClickListener {
            setCurrentTab(Constants.TAB_WALLET)
            get_free_coin_tv.visibility = View.GONE
        }
        firstCreateAccountActivity = Utils.getBoolean(Constants.FIST_CREATE_ACCOUNT, true)
        setCurrentTab(Constants.TAB_HOME)
        if (firstCreateAccountActivity) {
            showGuide()
            Utils.saveBoolean(Constants.FIST_CREATE_ACCOUNT, false)
        }
    }

    override fun initObserve() {
        mViewModel.hindeFreeCoinEvent.observe(this, Observer {
            if (it) {
                get_free_coin_tv.visibility = View.GONE
            } else {
                get_free_coin_tv.visibility = View.VISIBLE
            }
        })


    }


    override fun initVariableId(): Int = BR.viewModel


    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        if (HopApplication.instance.isRunning) {
            return
        }
        mViewModel.getWalletInfo(false)
    }


    fun loadWallet(isShowLoading: Boolean) {
        mViewModel.getWalletInfo(isShowLoading)
    }

    private fun initNavigator(savedInstanceState: Bundle?) {
        val bottomNavigatorAdapter = BottomNavigatorAdapter(this)
        for (fragment in mFragmentArray) {
            bottomNavigatorAdapter.addTab(BottomNavigatorAdapter.TabInfo(fragment.simpleName, fragment, null))
        }
        mNavigator = FragmentNavigator(supportFragmentManager, bottomNavigatorAdapter, R.id.content_frame)
        mNavigator.setDefaultPosition(Constants.TAB_HOME)
        mNavigator.onCreate(savedInstanceState)
    }

    override fun initData() {}
    private fun setCurrentTab(position: Int) {
        mNavigator.showFragment(position)
        if (null != main_bottom_navigator_view) {
            main_bottom_navigator_view.select(position)
        }
        setFreeCoinStatus(position)
    }

    private fun setFreeCoinStatus(position: Int) {
        if (position != Constants.TAB_WALLET && WalletWrapper.EthBalance == 0.0 && netTyp == Constants.TEST_NET) {
            get_free_coin_tv.visibility = View.VISIBLE
        } else {
            get_free_coin_tv.visibility = View.GONE

        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun addMiningPool(eventSkipTabPacketsMarket: EventSkipTabPacketsMarket) {
        setCurrentTab(Constants.TAB_RECHARGE)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mNavigator.onSaveInstanceState(outState)
    }


    @Subscribe(threadMode = ThreadMode.MAIN)
    fun rechargeSuccess(eventRechargeSuccess: EventRechargeSuccess) {
        mViewModel.syncSubPoolsData()
        loadWallet(false)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun eventReLoadWallet(event: EventReLoadWallet) {
        loadWallet(event.showDialog)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun showGuide(eventShowHomeTip: EventShowHomeTip) {
        setCurrentTab(Constants.TAB_HOME)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            val intent = Intent(Intent.ACTION_MAIN)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            intent.addCategory(Intent.CATEGORY_HOME)
            startActivity(intent)
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        super.onDestroy()
        WalletWrapper.closeWallet()
        if (!HopApplication.instance.isRunning) {
            AndroidLib.stopProtocol()
        }

        EventBus.getDefault().unregister(this)
    }

    fun showGuide() {
        higghtLight = HighLight(this).autoRemove(true).intercept(true).setClickCallback {
            setCurrentTab(Constants.TAB_WALLET)
        }.setOnLayoutCallback {
            higghtLight.addHighLight(R.id.bottom_navigator_wallet, R.layout.guide1, PirateOnTopPosCallback(), RectLightShape()).show()
        }.enableNext()
    }

}